docs/how-to-write-a-module.md

```diff
@@ -0,0 +1,252 @@
+
+### Charts method
+
+-   `Charts` returns the charts' definition.
+-   Called after `Check` and only if `Check` returned `true`.
+-   If it returns `nil`, the job will be disabled
+-   :warning: Make sure not to share returned value between module instances (jobs).
+
+Usually charts initialized in `Init` and `Chart` method just returns the charts instance:
+
+```go
+// example.go
+
+func (e *Example) Charts() *Charts {
+    return e.charts
+}
+```
+
+### Collect method
+
+-   `Collect` collects metrics.
+-    Called only if `Check` returned `true`.
+-    Called every `update_every` seconds.
+-    `map[string]int64` keys are charts dimensions ids'.
+
+We propose to use the following template:
+ 
+```go
+// example.go
+
+func (e *Example) Collect() map[string]int64 {
+    ms, err := e.collect()
+    if err != nil {
+        e.Error(err)
+    }
+
+    if len(ms) == 0 {
+        return nil
+    }
+    return ms
+}
+```
+
+Move metrics collection logic into the `collect.go` file. See [suggested module layout](#module-Layout).
+
+### Cleanup method
+
+-   `Cleanup` performs the job cleanup/teardown.
+-    Called if `Init` or `Check` fails, or we want to stop the job after `Collect`.
+
+
+If you have nothing to clean up:
+
+```go
+// example.go
+
+func (Example) Cleanup() {}
+```
+
+## Module Layout
+
+The general idea - do not put everything in a file.
+
+We recommend using a file per a logical area. This approach makes it easier to maintain the module.
+
+Suggested minimal layout:
+
+| Filename                                          | Contains                                               |
+| ------------------------------------------------- |------------------------------------------------------- |
+| [`module_name.go`](#file-module_namego)           | Module configuration, implementation and registration. |
+| [`charts.go`](#file-chartsgo)                     | Charts, charts templates and constructor functions.    |
+| [`init.go`](#file-initgo)                         | Initialization methods.                                |
+| [`collect.go`](#file-collectgo)                   | Metrics collection implementation.                     |
+| [`module_name_test.go`](#file-module_name_testgo) | Public methods/functions tests.                        |
+| [`testdata/`](#file-module_name_testgo)           | Tests fixtures.                                        |
+
+### File `module_name.go`
+
+> :exclamation: See the example [`examlpe.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/example/example.go).
+
+Don't overload this file with the implementation details.
+
+Usually it contains only:
+ 
+-   module registration.
+-   module configuration.
+-   [module interface implementation](#module-interface).
+
+### File `charts.go`
+
+> :exclamation: See the example: [`charts.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/example/charts.go).
+
+Put charts, charts templates and charts constructor functions in this file. 
```

Comment:
> See the example: [`charts.go`]

Reply:
I think the document should be all-encompassing, meaning that I don't need to go here and there to get required information to complete my understanding. Furthermore, I still can't understand what those data structure are or what is their purpose. 
 Let's get as explicit as possible, ELI5 level.

Reply:
Our goal is to lower the barrier of entry as much as possible, so that people can contribute easier and with less required technical understanding.

Reply:
We expect people checking other modules source code and source code in general (it is very easy to do useing IDE these days), it is impossobile ot avoid it.

You think that this is difficult, but it is not.

Reply:
I checked code you mean, but what I can't understand is what those data structures means for module, from a logical perspective.

The idea is to allow even rookie go programmers to easily grasp the fundamentals and create modules, otherwise, people will continue using python.

Reply:
I checked the file. I want us to explain to users what those mean in the netdata context, what is a chart and what is a template. Why in the example, one is hidden, what's the functionality of a hidden chart (i.e why would I want to do it). 
Give me all the information so that I don't need to make **any** assumption. Allow me to focus on business logic and now to understand the netdata aspects.